crystalfontz/dbus/domain/base.py

```diff
@@ -0,0 +1,213 @@
+from typing import (
+    ClassVar,
+    List,
+    Optional,
+    Protocol,
+    Self,
+    Type,
+    Union,
+)
+
+
+class TypeProtocol(Protocol):
+    t: ClassVar[str]
+
+
+def t(*args: Union[str, Type[TypeProtocol]]) -> str:
+    type_ = ""
+
+    for arg in args:
+        if isinstance(arg, str):
+            type_ += arg
+        else:
+            type_ += arg.t
+
+    return type_
+
+
+def struct(*args: Union[str, Type[TypeProtocol]]) -> str:
+    return t("(", *args, ")")
+
+
+def array(of: Union[str, Type[TypeProtocol]]) -> str:
+    return f"a{t(of)}"
+
+
+class NoneM:
+    t: ClassVar[str] = ""
+
+
+OptUintT = int
+
+
+class OptUintM:
+    """
```

Comment:
Now that I've made dbus types more explicit, I think I should revisit these docstrings...

Reply:
I'll do this in #75.